src/cls/queue/cls_queue_ops.h

```diff
@@ -0,0 +1,204 @@
+#ifndef CEPH_CLS_QUEUE_OPS_H
+#define CEPH_CLS_QUEUE_OPS_H
+
+#include "cls/queue/cls_queue_types.h"
+
+struct cls_create_queue_op {
+  cls_queue_head head;
+  uint64_t head_size{0};
+
+  cls_create_queue_op() {}
+
+  void encode(bufferlist& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(head, bl);
+    encode(head_size, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator& bl) {
+    DECODE_START(1, bl);
+    decode(head, bl);
+    decode(head_size, bl);
+    DECODE_FINISH(bl);
+  }
+
+  void dump(Formatter *f) const;
+  static void generate_test_instances(list<cls_create_queue_op*>& ls);
+};
+WRITE_CLASS_ENCODER(cls_create_queue_op)
+
+struct cls_enqueue_op {
+  vector<bufferlist> bl_data_vec;
+  bufferlist bl_urgent_data;
+
+  cls_enqueue_op() {}
+
+  void encode(bufferlist& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(bl_data_vec, bl);
+    encode(bl_urgent_data, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator& bl) {
+    DECODE_START(1, bl);
+    decode(bl_data_vec, bl);
+    decode(bl_urgent_data, bl);
+    DECODE_FINISH(bl);
+  } 
+};
+WRITE_CLASS_ENCODER(cls_enqueue_op)
+
+struct cls_dequeue_op {
+  bufferlist bl;
+
+  cls_dequeue_op() {}
+
+  static void generate_test_instances(list<cls_dequeue_op*>& ls);
+};
+
+struct cls_queue_list_op {
+  uint64_t max;
+  uint64_t start_offset;
+
+  cls_queue_list_op() {}
+
+  void encode(bufferlist& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(max, bl);
+    encode(start_offset, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator& bl) {
+    DECODE_START(1, bl);
+    decode(max, bl);
+    decode(start_offset, bl);
+    DECODE_FINISH(bl);
+  }
+};
+WRITE_CLASS_ENCODER(cls_queue_list_op)
+
+struct cls_queue_list_ret {
+  bool is_truncated;
+  uint64_t next_offset;
+  vector<bufferlist> data;
+  vector<uint64_t> offsets;
```

Comment:
Can you explain a little more on how the gen number will help in cls_queue_remove_entries(). The wrapping around can be deduced by the front and the tail offsets.

It makes sense to have opaque offsets returned back to clients for list entries. and also to have a constant at the beginning of a valid queue entry, because I did encounter problems where an invalid 8 byte was decoded and that crashed the osd while trying to read data using the invalid length.

Reply:
> Can you explain a little more on how the gen number will help in cls_queue_remove_entries(). The wrapping around can be deduced by the front and the tail offsets.

consider the case where you completely fill the queue such that `head.front == head.tail`, and get a remove_entries() op with that same offset. how can you tell whether the request wants to trim all entries up to head.tail, or if it was only supposed to trim up to head.front?

Reply:
But why will an application call a remove with the same offset as head.front == head.tail, when it doesn't want to trim any entry? And in case if it wants to trim the head.front entry, it will use the start offset of last+1 entry as end offset to remove entries.

Reply:
@cbodley : We need to take care of integer overflow of gen number right? reset it back to zero when it reaches its maximum value?

Reply:
Yes, it will need to roll over.